src/nix.rs

```diff
@@ -339,6 +337,24 @@ impl CallOpts {
     }
 }
 
+/// Helper function to correctly parse the output of a `std::process::Output`’s
+/// `stdout` and `stderr` for nix outputs.
+pub fn parse_nix_output<'a, T, F>(output: &'a [u8], f: F) -> Vec<T>
+where
+    F: Fn(&'a OsStr) -> T,
+{
+    output
+        // We can split on \n to separate lines, because nix only runs in POSIX environments.
+        // Using `lines()` means having to convert to UTF-8 first,
+        // and nix output is not guaranteed to be valid UTF-8.
+        // (think for example a derivation that outputs random data).
+        .split(|b| &b'\n' == b)
+        .map(std::ffi::OsStr::from_bytes)
+        .filter(|s| !s.is_empty())
```

Comment:
Interesting! I think it'd be good to make a module, with a `Lines<OsStr>` iterator, and a function like:

```
fn lines(&OsStr) -> Lines<OsStr>
```

though maybe accepting a `&'a [u8]` or even `OsString` is okay too. Not sure if there is one to be preferred.

Then, the `.map().collect()` can be part of a different function, or an iter chain.

_note: don't filter out empty lines. empty lines are lines too :)_

---

After this change, the call sites of `parse_nix_output` are (mostly) just `lines(output).map(...).collect()`

Reply:
I’ll leave that for a future improvement, when we really need streaming.

Reply:
Was provided by the improvements @grahamc made.